c_common.c

```diff
@@ -98,30 +101,31 @@ uint16_t read_variable_length_one_byte_each(int socket, unsigned char *buf) {
 }
 
 int read_header_return_data_buf_pointer(int socket, unsigned char *message_type,
-                                        unsigned char *ret,
-                                        unsigned int *ret_length) {
+                                        unsigned char *buf,
+                                        unsigned int buf_length) {
     unsigned char received_buf[MAX_PAYLOAD_BUF_SIZE];
-    int socket_read = read(socket, received_buf, MESSAGE_TYPE_SIZE);
-    if (socket_read == 0) {
-        printf("Socket closed!\n");
-        close(socket);
-        return 0;
-    }
-    if (socket_read == -1) {
-        printf("Connection error!\n");
-        return 0;
-    }
+    // Read the first byte.
+    read_from_socket(socket, received_buf, MESSAGE_TYPE_SIZE);
     *message_type = received_buf[0];
+    // Read one bytes each, until the variable length buffer ends.
     unsigned int var_length_buf_size = read_variable_length_one_byte_each(
         socket, received_buf + MESSAGE_TYPE_SIZE);
     unsigned int var_length_buf_size_checked;
+    unsigned int ret_length;
+    // Decode the variable length buffer and get the bytes to read.
     var_length_int_to_num(received_buf + MESSAGE_TYPE_SIZE, var_length_buf_size,
-                          ret_length, &var_length_buf_size_checked);
+                          &ret_length, &var_length_buf_size_checked);
     if (var_length_buf_size != var_length_buf_size_checked) {
         error_exit("Wrong header calculation... Exiting...");
     }
-    read(socket, ret, *ret_length);
-    return 1;
+    if (ret_length > buf_length) {
+        error_exit("Larger buffer size required.");
+    }
+    unsigned int bytes_read = read_from_socket(socket, buf, buf_length);
+    if (ret_length != bytes_read) {
+        error_exit("Wrong read... Exiting..");
+    }
+    return bytes_read;
```

Comment:
Same here, it is clearly not reformatting.

Reply:
I updated the PR description. Thank you for having a look.